### PR TITLE
feat: `Value` representation of `Simple` values

### PR DESCRIPTION
**Pull request by emonadeo**
Currently Simple values are not representable as a Value type.

We (@emonadeo, @hw0lff, @TheodorStraube, @moehr1z) are working on a [Packed CBOR] implementation based on ciborium, which relies on tags and simple values.

This PR adds a `Simple` variant to the `Value` enum.

### Changes

The changes basically boil down to the following:

```diff
pub enum Value {
    Integer(Integer),
    Bytes(Vec<u8>),
    Float(f64),
    Text(String),
    Bool(bool),
    Null,
    Tag(u64, Box<Value>),
    Array(Vec<Value>),
    Map(Vec<(Value, Value)>),
+   Simple(u8),
}
```

In order to serialize `Value::Simple(u8)`, `serialize_newtype_struct` and hardcoded `@@SIMPLE@@` markers are used (similar to how it is done with Tags)

This works, but is admittedly not great. Potential other ideas here are appreciated.

### RFC: Forbidden two-byte encodings of one-byte simple values:

[4 tests](https://github.com/emonadeo/ciborium/blob/a4a9170da587630723189bd049b1808c06b4169d/ciborium/tests/error.rs#L117-L121) are failing at the moment:

```rust
// Forbidden two-byte encodings of one-byte simple values:
case("f800", Error::Semantic(None, "...".into())),
case("f801", Error::Semantic(None, "...".into())),
case("f818", Error::Semantic(None, "...".into())),
case("f81f", Error::Semantic(None, "...".into())),
```

These are testing `simple(0)`, `simple(1)`, `simple(24)` and `simple(31)` respectively.
While syntactically correct, the CBOR Specification states:

> An encoder MUST NOT issue two-byte sequences that start with 0xf8 (major type 7, additional information 24) and continue with a byte less than 0x20 (32 decimal).[^1]

(This also means that `simple(24)` can never exist, as `f818` is not valid and `f8` by itself denotes the use of an additional byte. Likewise `simple(25)`..`simple(31)` cannot exist, because they are either floats, reserved, or a break stop code).

We intentionally left this unfixed, as we are not sure if this should be caught in the decoder of ciborium_ll, or at a higher level in ciborium.

### RFC: Replace `Bool` and `Null` variants with `Simple` constants

We originally removed the `Bool` and `Null` variants and replaced them with constants, since these are technically also `Simple` values:

```diff
pub enum Value {
    Integer(Integer),
    Bytes(Vec<u8>),
    Float(f64),
    Text(String),
-   Bool(bool),
-   Null,
    Tag(u64, Box<Value>),
    Array(Vec<Value>),
    Map(Vec<(Value, Value)>),
+   Simple(u8),
}

+ pub const FALSE: Value = Value::Simple(ciborium_ll::simple::FALSE);
+ pub const TRUE: Value = Value::Simple(ciborium_ll::simple::TRUE);
+ pub const NULL: Value = Value::Simple(ciborium_ll::simple::NULL);
+ pub const UNDEFINED: Value = Value::Simple(ciborium_ll::simple::UNDEFINED);
```

However removing these would introduce a breaking change. As such we figured a redundant representation of `false` _(`Value::Bool(false)` or `Value::Simple(20)`)_, `true` _(`Value::Bool(true)` or `Value::Simple(21)`)_ and `null` _(`Value::Null` or `Value::Simple(22)`)_ makes more sense in order to both not break compatibility and preserve ergonomics.

[Packed CBOR]: https://datatracker.ietf.org/doc/draft-ietf-cbor-packed/

[^1]: https://www.rfc-editor.org/rfc/rfc8949.html#section-3.3-5

### CODE REVIEW

File: ciborium/src/value/canonical.rs [hunk 1]
```diff
@@ -46,15 +46,14 @@ pub fn cmp_value(v1: &Value, v2: &Value) -> Ordering {
             Ordering::Equal => s.cmp(o),
             x => x,
         },
-        (Bool(s), Bool(o)) => s.cmp(o),
-        (Null, Null) => Ordering::Equal,
```
Comment (rjzak) [inline ciborium/src/value/canonical.rs:50]: Do these items need to be removed?

File: ciborium/src/simple.rs [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+
+impl ser::StdError for Error {}
+
+impl ser::Error for Error {
+    fn custom<U: core::fmt::Display>(_msg: U) -> Self {
+        Error
+    }
+}
+
+pub(crate) struct Serializer;
```
Comment (rjzak) [inline ciborium/src/simple.rs:81]: Why are all the functions for `Serializer` returning an error?

**Review by rjzak — COMMENTED** [review]
Is this expected to change the API from the user's perspective?

**Comment by rjzak** [comment]
> We intentionally left this unfixed, as we are not sure if this should be caught in the decoder of ciborium_ll, or at a higher level in ciborium.

I would think that's something for ciborium_ll.

File: ciborium/src/value/canonical.rs (hunk 1 above)
Comment (emonadeo) [reply ciborium/src/value/canonical.rs:50]: Good catch, I don't think they do. I assume this an artifact of our first implementation, that removed `Value::Bool` and `Value::Null`, which we didn't properly undo.

File: ciborium/src/simple.rs (hunk 2 above)
Comment (emonadeo) [reply ciborium/src/simple.rs:81]: The only function returning `Ok` is `serialize_u8`, which is the only legal content of a simple value. It is used in [ciborium/src/ser/mod.rs (`Serializer::serialize_newtype_struct`)](https://github.com/enarx/ciborium/pull/137/commits/19fafbac0d25b42a6f1ef50fcc329da9ca0b5057#diff-bc9bec000d361bc167ed6da072a6726893eea63993e0e4074a8fd879c6389b96R207)

**Comment by emonadeo** [comment]
> Is this expected to change the API from the user's perspective?

This PR? No.

The WIP Packed CBOR impl? Not at the moment. At its current state we simply added new functions for packed cbor.

**Comment by rjzak** [comment]
A few of the `ciborium` tests aren't passing.

**Comment by emonadeo** [comment]
> A few of the `ciborium` tests aren't passing.

You mean any tests beyond the four mentioned in the PR?

**Comment by rjzak** [comment]
> > A few of the `ciborium` tests aren't passing.
> 
> You mean any tests beyond the four mentioned in the PR?

No, sorry; I forgot about those four.

**Comment by npmccallum** [comment]
Having simple values as a separate Value variant is a major strategic direction change for this crate. I'm not sure it makes sense.

**Comment by npmccallum** [comment]
@emonadeo I really don't understand what you're suggesting here. `ciborium` always chooses the most compact representation. Why is this change necessary? What byte stream can't you produce/parse today?

**Comment by npmccallum** [comment]
@emonadeo But what is the semantic meaning of `Simple(0)`? The `Value` struct should only contain semantically interpreted values. For example: I would accept `Value::Undefined` since it has a semantic value. But `Simple(23)` has no semantic value.

**Comment by emonadeo** [comment]
> The Value struct should only contain semantically interpreted values.

Oh, do they? Tags also aren't semantic and they have a Value representation.

> But what is the semantic meaning of Simple(0)?

It doesn't have an immediate semantic value. Packed CBOR uses `Simple(0)` as an intermediary value to reference other values.

I don't think it's possible to implement Packed CBOR on top of Ciborium without a way to deserialize Simple values. Adding representation of Simple values opens up more possibilities to use Ciborium for „extension protocols“ as with the case of Packed CBOR.

**Comment by npmccallum** [comment]
> > The Value struct should only contain semantically interpreted values.
> 
> Oh, do they? Tags also aren't semantic and they have a Value representation.

Tags are at least semi-semantic. They contain a hint as to the proper usage of the following type.

`Simple(0)` on the other hand contains no semantic content.

> > But what is the semantic meaning of Simple(0)?
> 
> It doesn't have an immediate semantic value. Packed CBOR uses `Simple(0)` as an intermediary value to reference other values.
> 
> I don't think it's possible to implement Packed CBOR on top of Ciborium without a way to deserialize Simple values. Adding representation of Simple values opens up more possibilities to use Ciborium for „extension protocols“ as with the case of Packed CBOR.

I'll be really honest, when I read the Packed CBOR RFC, it very much strikes me as a solution in search of a problem. It has only two authors, both in academia. It solves a problem that can be easily solved by just gzipping the cbor. It solves the problem in a way that significantly complicates encoding. And it solves the problem less efficiently than a good compression algorithm. Here's the main paragraph:

```
CBOR does not provide any forms of data compression.  CBOR data
items, in particular when generated from legacy data models, often
allow considerable gains in compactness when applying data
compression.  While traditional data compression techniques such as
DEFLATE [RFC1951] can work well for CBOR encoded data items, their
disadvantage is that the recipient needs to decompress the compressed
form to make use of the data.
```

Which use cases are hesitant to decompress the CBOR before using it? The industry has reliable, battle-hardened and highly optimized compression algorithms and implementations. And further, they can be implemented at the transport layer, so applications don't need to even think about this detail.

I just don't "get it". Perhaps it is because I'm dense.

Can you give me a concrete case where packed CBOR is an essential feature?

**Comment by npmccallum** [comment]
@emonadeo I definitely don't want to expose `Simple` in the `Value` struct. That's mixing low-level encoding details with high-level semantic details.

I spent some time reading on the packed CBOR proposal today. My findings aren't really positive.

1. There is [this very reasonable critique](https://mailarchive.ietf.org/arch/msg/cbor/qdMZwu-CxHT5XP0njjGkBSi2pVQ/) that has been completely ignored by the authors.

2. Almost all of the activity on this topic appears to be by the draft authors. This is a huge red flag.

3. The cost of this proposal outweighs the benefits. Serialization and deserialization become much more complicated. Compression is less efficient. Applications have to be involved in the compression, so there is no clear layer separation. The amount of the tags and simple values consumed by this single proposal is astronomically high. I just can't see that the benefit is proportional to the cost.

That being said, I'm willing to see a patch against the serializer and deserializer to enable a "packed" mode. I'd like to see how invasive it is. The application would need to opt-in to packed mode. Likewise, I'd want packed mode behind a feature flag so that we can rip it out.

But it is not a goal of the ciborium crate to enable a separate crate to implement packed mode. And we are definitely not going to add `Simple` to the `Value` struct.

**Pull request closed without merging**